### PR TITLE
Hide the Shelley update-epoch transition logic behind a UPEC rule.

### PR DESCRIPTION
**Pull request by dnadales**
This will allow us to easily change this aspect from era to era.

### CODE REVIEW

**Comment by dnadales** [comment]
Problems I see with the current PR:
- The UPEC state is the `EpochState`, which contains the entire ledger state!
- Shelley assumes that PPUP and therefore UPEC will update deposits and
  reserves. So if another update implementation does not do this, things will break!

A solution might be taking the update of the deposits and reserves away
from `PPUP` and into the `EPOCH` rule. This would mean the the `PPUP`
might need to decide whether to update the protocol parameters, but it
will be the `EPOCH` rule the one that will update the monetary aspects
of the humongous `EpochState`.

**Comment by JaredCorduan** [comment]
I think it's fine. it's definitely unpleasant that it has no environment and no signal, and nearly all the ledger state. but, as you know, this is all just coming from the fact that we don't track what the deposit values where at different points in time. if we did decide to start tracking them (which is probably the right thing to do if we aren't worried about the size), then we can make those disadvantages go away. if you are happy with it, let's do it, and hope that we can get to tracking individual deposit values.

**Comment by dnadales** [comment]
> I think it's fine. it's definitely unpleasant that it has no environment and no signal, and nearly all the ledger state. but, as you know, this is all just coming from the fact that we don't track what the deposit values where at different points in time. if we did decide to start tracking them (which is probably the right thing to do if we aren't worried about the size), then we can make those disadvantages go away. if you are happy with it, let's do it, and hope that we can get to tracking individual deposit values.

I'm still not happy with it :) Tomorrow I'd like to try to move the monetary logic from PPUP to EPOCH, which cause the UPEC rule to take the epoch state as environment, but modify only the update state and protocol parameters. In that way Shelley does not have the implicit assumption that the UPEC rule will update the deposits and reserves. If you think this does not make sense, then I won't even bother :)

**Comment by JaredCorduan** [comment]
I like this a lot better @dnadales , thank you!

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Epoch.hs [hunk 1]
```diff
@@ -149,27 +128,39 @@ epochTransition = do
...
+
+  UPECState pp' ppupSt' <-
+    trans @(UPEC era) $ TRC (epochState', UPECState pp (_ppups utxoSt'), ())
+  let utxoSt'' = utxoSt' {_ppups = ppupSt'}
+
+  let Coin oblgCurr = obligation pp (_rewards dstate') (_pParams pstate'')
+      Coin oblgNew = obligation pp' (_rewards dstate') (_pParams pstate'')
+      Coin reserves = _reserves acnt'
+      utxoSt''' =
+        assert (_deposited utxoSt'' == Coin oblgCurr) $
```
Comment (dnadales) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Epoch.hs]: I'll remove this assertion. I just wanted to check that it is true. @JaredCorduan do you know if this is always the case?

Comment (JaredCorduan) [reply shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Epoch.hs]: on the epoch boundary, yes, we expect that to be true

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Upec.hs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  }
+  deriving (Show)
+
+data UpecPredicateFailure era
+  = NewPpFailure (PredicateFailure (NEWPP era))
+  deriving (Eq, Show, Generic)
+
+instance NoThunks (UpecPredicateFailure era)
+
+instance ShelleyBased era => STS (UPEC era) where
```
Comment (nc6) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Upec.hs]: We're trying to drop `ShelleyBased` in favour of more fine-grained constraints

**Review by nc6 — COMMENTED** [review]
Looks reasonable - I've left one comment. Since this is a big enough change, we need to ensure it doesn't break validation of the existing chain before merging. I can do this if you like,

**Review by nc6 — APPROVED** [review]
Tested up to the latest update (Allegra HF). Everything looks tickety-boo. Thanks @dnadales!

**Pull request merged**